### DISCUSSION TITLE
[Feature] Sphere/Panorama view for the mobile client

### DISCUSSION THREAD

**Discussion by NtWriteCode** [immich-app/immich#6572] [Feature Request]
### The feature

It would be awesome if the mobile client could support viewing sphere or panorama views made by our camera :)

I've already seen that the webUI supports it, however when I just want to see what photo I took on my device, it's a bit sad that I have to use a third party app (Google Photos) to see the result.

### Platform

- [ ] Server
- [ ] Web
- [X] Mobile

**Comment by iansewell** [comment]
and bonus points if we can get 360 videos working! (mp4), like google photos.  But lower prio than 360 photos on mobile..

**Reply by DeltaTango69** [reply]
Yes that would be awesome. I have tons of videos from my Insta x3 and EVO :-)

**Comment by Shurov** [comment]
I'm so few people are interested (judging by the up emojis and comments #). So, vouching +1!!!!

**Comment by FreeWind6** [comment]
Viewing photos and videos in 360 format on a mobile device is very much lacking =(

**Reply by DeltaTango69** [reply]
When did this start working? My 360° photos are only displayed flat in the app. I can't test videos because the transcoded videos are not displayed.

**Reply by FreeWind6** [reply]
Photos and videos 360 displayed only in the web version. There is no support for 360 photos and videos in the mobile app

**Reply by DeltaTango69** [reply]
Yes, I know, so you meant the web version on the mobile and not the app - now I get it  :-)

**Comment by freeKC** [comment]
Immich is such an amazing solution overall. I particularly miss the gyroscope-enabled 360° viewing feature on mobile - being able to naturally pan around by moving the phone would make the viewing experience so much more intuitive. Can't wait to see this implemented in the mobile app! 🙌

**Comment by CustomIcon** [comment]
very much requested feature for my use case as well. i take 360 photos of places i visit using my dji drone and view them in the browser. the app lacks this feature sadly

**Comment by Smarti15** [comment]
Please add support for 360° images and, later on, videos as well.

**Comment by Matou25** [comment]
360° photos in mobile app will be a must, all insta 360 and other 360 camera owners will adopt immich

**Comment by Snufkin-8** [comment]
May I ask any updates for this, any chance to have this feature implemented?

**Reply by bo0tzz** [reply]
https://justinmayer.com/posts/any-updates/

**Comment by ChazyChaxxx** [comment]
I'm interessed too 🙏

**Comment by bencefr** [comment]
I'm not a flutter developer and I don't dare to make a PR with AI generated junk, but as a concept I have tried to use [panorama_viewer](https://pub.dev/packages/panorama_viewer),  and it works quite nicely even with images that are 18714x9357 39.4 MiB jpegs. My quick implementation requires the images to be local assets - which is fine - and it performs way better than the web based viewer.

Some snippets:
[mobile/lib/presentation/widgets/asset_viewer/asset_page.widget.dart](https://github.com/immich-app/immich/blob/c9539a4aa8814b2d05ab267386d90cc1842b1c9c/mobile/lib/presentation/widgets/asset_viewer/asset_page.widget.dart#L340)
```dart
    ...
    if (asset.isImage && !isPlayingMotionVideo) {
      final exifInfo = ref.watch(assetExifProvider(asset)).valueOrNull;
      final isPhotosphere = exifInfo?.projectionType?.toUpperCase() == 'EQUIRECTANGULAR' || asset.isPhotosphere;
      if (isPhotosphere && !showingControls && !showingDetails) {
        return PanoramaViewer(
          interactive: true,
          sensorControl: SensorControl.none,
          onTap: (lat, lon, angle) => _viewer.toggleControls(),
          child: Image(
            image: imageProvider,
            fit: BoxFit.cover,
            gaplessPlayback: true,
            filterQuality: FilterQuality.high,
          ),
        );
      }
    ...
```
_showingControls_ and _showingDetails_ must be propagated and the idea is that the panorama viewer is only loaded when the controls (and details) are hidden.

I don't know how and where to best handle the projectionType properly for remote and local assets, but the above check somehow gets it from exif. I also added `isPhotoshpere` to base assets:
[mobile/lib/domain/models/asset/base_asset.model.dart](https://github.com/immich-app/immich/blob/c9539a4aa8814b2d05ab267386d90cc1842b1c9c/mobile/lib/domain/models/asset/base_asset.model.dart)
```dart
  bool get isPhotosphere =>
      isImage && hasLocal && width != null && height != null && height! > 0 && (width! / height!) >= 1.99;
```

For indication in the thumbnail view that the image is a photosphere I added the icon:
[mobile/lib/presentation/widgets/images/thumbnail_tile.widget.dart](https://github.com/immich-app/immich/blob/c9539a4aa8814b2d05ab267386d90cc1842b1c9c/mobile/lib/presentation/widgets/images/thumbnail_tile.widget.dart#L315)
```dart
        if (asset.isPhotosphere)
          const Padding(
            padding: EdgeInsets.only(right: 10.0, top: 6.0),
            child: _TileOverlayIcon(Icons.panorama_photosphere),
          ),
```

It would be nice to add controls to the panorama viewer so e.g. sensor control could be toggled.

Someone fluent in this ecosystem, please turn this into a round PR.

**Comment by CoderXL1** [comment]
I second that. It's been 2.5 years btw, and it's still not added. :=(
